### PR TITLE
Refactor Backend: Merge UserSQL and User

### PR DESCRIPTION
**Pull request by fm3**
### URL of deployed dev instance (used for testing):
- https://refactorbackenduser.webknossos.xyz

### Steps to test:
- set up wK with multiple user accounts (try registration form also)
- check that rights/team management still works
- click around, general testing
- user viewport/dataset settings should work (be saved, loaded again after refresh)
- create some tasks, trace some, download project

### Issues:
- contributes to #2521

------
- ~~[ ] Updated [changelog](../blob/master/CHANGELOG.md#unreleased)~~
- ~~[ ] Updated [migration guide](../blob/master/MIGRATIONS.md#unreleased) if applicable~~
- [x] Ready for review

### CODE REVIEW

File: app/controllers/AnnotationIOController.scala [hunk 1]
```diff
@@ -170,20 +170,19 @@ class AnnotationIOController @Inject()(val messagesApi: MessagesApi)
...
   }
 
-  def downloadProject(projectId: String, user: User)(implicit ctx: DBAccessContext) = {
+  def downloadProject(projectId: String, user: UserSQL)(implicit ctx: DBAccessContext) = {
     for {
       projectIdValidated <- ObjectId.parse(projectId)
       project <- ProjectSQLDAO.findOne(projectIdValidated) ?~> Messages("project.notFound", projectId)
-      teamIdBson <- project._team.toBSONObjectId.toFox
-      _ <- user.assertTeamManagerOrAdminOf(teamIdBson)
+      _ <- Fox.assertBoolean(user.isTeamManagerOrAdminOf(project._team))
```
Comment (jstriebel) [inline app/controllers/AnnotationIOController.scala]: This purely waits for the future and checks that it is not `None`, right? Seems to be the correct refactoring, still wondering what this line does.

File: app/controllers/ConfigurationController.scala [hunk 2]
```diff
@@ -1,27 +1,25 @@
...
- * Controller that handles the CRUD api for configurations (mostly settings for the tracing view)
- */
+
 class ConfigurationController @Inject()(val messagesApi: MessagesApi) extends Controller {
 
   def read = UserAwareAction.async { implicit request =>
     request.identity.toFox.flatMap { user =>
-      UserService.findOneById(user.id, useCache = false)
-      .map(_.userConfiguration.configurationOrDefaults)
+      for {
```
Comment (jstriebel) [inline app/controllers/ConfigurationController.scala]: just out of curiosity: why did you switch here to the `for` syntax?

File: app/controllers/ConfigurationController.scala [hunk 3]
```diff
@@ -39,8 +37,9 @@ class ConfigurationController @Inject()(val messagesApi: MessagesApi) extends Co
 
   def readDataSet(dataSetName: String) = UserAwareAction.async { implicit request =>
     request.identity.toFox.flatMap { user =>
-      UserService.findOneById(user.id, useCache = false)
-      .flatMap(_.dataSetConfigurations.get(dataSetName))
+      for {
```
Comment (jstriebel) [inline app/controllers/ConfigurationController.scala]: This seems to be more than a refactoring. If I understand it correctly, this used to be cached in the `UserCache`, not sure if it was used from there.

And for my understanding: `findOneForUserAndDataset` comes from slick now, right?

File: app/controllers/Controller.scala [hunk 4]
```diff
@@ -31,31 +31,25 @@ trait Controller extends PlayController
   implicit def AuthenticatedRequest2Request[T](r: SecuredRequest[T]): Request[T] =
     r.request
 
-  def ensureTeamAdministration(user: User, teamId: ObjectId): Fox[Unit] =
-    for {
-      teamIdBson <- teamId.toBSONObjectId.toFox
-      _ <- ensureTeamAdministration(user, teamIdBson)
-    } yield ()
+  def ensureTeamAdministration(user: UserSQL, teamId: ObjectId): Fox[Unit] =
+    Fox.assertBoolean(user.isTeamManagerOrAdminOf(teamId)) ?~> Messages("team.admin.notAllowed")
```
Comment (jstriebel) [inline app/controllers/Controller.scala]: Does `assertBoolean` actually assert it's true? It looks like it should, but it doesn't sound like it does.

File: app/controllers/ConfigurationController.scala [hunk 5]
```diff
@@ -68,7 +67,8 @@ class ConfigurationController @Inject()(val messagesApi: MessagesApi) extends Co
   def updateDataSetDefault(dataSetName: String) = SecuredAction.async(parse.json(maxLength = 20480)) { implicit request =>
     for {
       dataset <- DataSetDAO.findOneBySourceName(dataSetName) ?~> Messages("dataset.notFound")
-      _ <- (request.identity.isAdminOf(dataset.owningOrganization) || request.identity.isTeamManagerInOrg(dataset.owningOrganization)) ?~> Messages("notAllowed")
+      organization <- OrganizationSQLDAO.findOneByName(dataset.owningOrganization)
+      _ <- Fox.assertBoolean(request.identity.isTeamManagerOrAdminOfOrg(organization._id)) ?~> Messages("notAllowed")
```
Comment (jstriebel) [inline app/controllers/ConfigurationController.scala]: Could we just use `ensureTeamAdministration` instead?

File: app/controllers/DataSetController.scala [hunk 6]
```diff
@@ -78,14 +78,15 @@ class DataSetController @Inject()(val messagesApi: MessagesApi) extends Controll
   def list = UserAwareAction.async { implicit request =>
     UsingFilters(
       Filter("isEditable", (value: Boolean, el: DataSet) =>
-        el.isEditableBy(request.identity) && value || !el.isEditableBy(request.identity) && !value),
+        for {isEditable <- el.isEditableBy(request.identity)} yield {isEditable && value || !isEditable && !value}),
```
Comment (jstriebel) [inline app/controllers/DataSetController.scala]: also here, why the `for` syntax?

File: app/controllers/DataSetController.scala [hunk 7]
```diff
@@ -78,14 +78,15 @@ class DataSetController @Inject()(val messagesApi: MessagesApi) extends Controll
...
       Filter("isActive", (value: Boolean, el: DataSet) =>
-        el.isActive == value)
+        Fox.successful(el.isActive == value))
     ) { filter =>
       DataSetDAO.findAll.flatMap {
         dataSets =>
           for {
-            js <- Fox.serialCombined(filter.applyOn(dataSets))(d => DataSet.dataSetPublicWrites(d, request.identity))
+            filtered <- filter.applyOn(dataSets)
```
Comment (jstriebel) [inline app/controllers/DataSetController.scala]: :heart:

File: app/controllers/DataSetController.scala [hunk 8]
```diff
@@ -176,7 +178,7 @@ class DataSetController @Inject()(val messagesApi: MessagesApi) extends Controll
       case (server, name, token, team) =>
         for {
           _ <- DataSetService.checkIfNewDataSetName(name) ?~> Messages("dataSet.name.alreadyTaken")
-          _ <- ensureTeamAdministration(request.identity, team)
+          _ <- Fox.assertBoolean(request.identity.isTeamManagerOrAdminOf(ObjectId.fromBsonId(team))) ?~> Messages("team.admin.notAllowed")
```
Comment (jstriebel) [inline app/controllers/DataSetController.scala]: Why not stay with the new `ensureTeamAdministration`? Seems to be equivalent.

File: app/controllers/AnnotationIOController.scala (hunk 1 above)
Comment (fm3) [reply app/controllers/AnnotationIOController.scala]: This is about functions that return `Fox[Boolean]`. At several locations, we want to check such booleans which we can only get packed in a Fox. One previous method to deal with that by creating helper functions for each, such as `assertTeamManagerOrAdminOf`, which unpacks the Fox returned by `isTeamManagerOrAdminOf`, and then checks the `content` of that boolean, by casting it to a new Fox via `bool2fox`. In this PR the number of these helper functions would have increased, so I refactored this into the helper function `Fox.assertBoolean`. It returns `Fox[Unit]`, which is full __only__ when the original function returns `Fox.successful(true)´.

File: app/controllers/ConfigurationController.scala (hunk 2 above)
Comment (fm3) [reply app/controllers/ConfigurationController.scala]: userConfigurationStructured now returns a Fox instead of the content directly (this is a common change in this PR). I tend to get confused with nested `flatMaps`, which would have now been necessary here.

File: app/controllers/ConfigurationController.scala (hunk 3 above)
Comment (fm3) [reply app/controllers/ConfigurationController.scala]: Correct observation, I did not consider that this is now bypassing the Cache. However, I believe that it is not problematic. This configuration is requested via a separate route (not part of the user json) and that route is requested far less often. Also, the UserCache’s cache invalidation should provide that there is not actually a change in behavior (albeit maybe a tiny one in performance)

File: app/controllers/ConfigurationController.scala (hunk 5 above)
Comment (fm3) [reply app/controllers/ConfigurationController.scala]: Yes, will do.

File: app/controllers/Controller.scala (hunk 4 above)
Comment (fm3) [reply app/controllers/Controller.scala]: I sure hope it does (see above)

File: app/controllers/DataSetController.scala (hunk 6 above)
Comment (fm3) [reply app/controllers/DataSetController.scala]: isEditableBy is now also returning a Fox, so the alternative to the `for-yield` would have been a `map` of some kind, and I find the `for` easier to read

File: app/controllers/DataSetController.scala (hunk 7 above)
Comment (fm3) [reply app/controllers/DataSetController.scala]: `filter.applyOn` is now also asynchronous so the result has to be unpacked, which is what this line does :)

File: app/controllers/DataSetController.scala (hunk 8 above)
Comment (fm3) [reply app/controllers/DataSetController.scala]: Yes, will do.
(the explanation is that I first wanted to remove the `ensureTeamAdministration` entirely because I found it a little superfluos and the name a little confusing, but then I saw that it was used in so many places that I stopped removing it. Also, I guess does help reduce clutter a little bit)

File: app/controllers/AnnotationIOController.scala (hunk 1 above)
Comment (jstriebel) [reply app/controllers/AnnotationIOController.scala]: Ah, ok. How about renaming it to `Fox.assertTrue`?

File: app/controllers/ConfigurationController.scala (hunk 2 above)
Comment (jstriebel) [reply app/controllers/ConfigurationController.scala]: got it, thanks :+1:

File: app/controllers/ConfigurationController.scala (hunk 3 above)
Comment (jstriebel) [reply app/controllers/ConfigurationController.scala]: Fine for me :+1:

File: app/controllers/UserController.scala [hunk 9]
```diff
@@ -28,22 +28,24 @@ class UserController @Inject()(val messagesApi: MessagesApi)
 
   val defaultAnnotationLimit = 1000
 
-  def current = SecuredAction { implicit request =>
-    Ok(Json.toJson(request.identity)(User.userPublicWrites(request.identity)))
+  def current = SecuredAction.async { implicit request =>
+    for {
```
Comment (jstriebel) [inline app/controllers/UserController.scala]: I think we have this pattern quite often, should we extract this at some point? (Not part of this PR, though.)

Comment (jstriebel) [reply app/controllers/UserController.scala]:
```
implicit request =>
for {
  …
  js <- something.publicWrites(request.identity)
} yield Ok(userJs)
```

File: app/controllers/UserController.scala [hunk 10]
```diff
@@ -61,8 +63,9 @@ class UserController @Inject()(val messagesApi: MessagesApi)
 
   def userLoggedTime(userId: String) = SecuredAction.async { implicit request =>
     for {
-      user <- UserDAO.findOneById(userId) ?~> Messages("user.notFound")
-      _ <- user.isEditableBy(request.identity) ?~> Messages("notAllowed")
```
Comment (jstriebel) [inline app/controllers/UserController.scala]: How was `isEditableBy` actually checked? Did this happen via the `?~>`? I thought that was only checking that the option is defined, but not the packed boolean.

File: app/controllers/UserController.scala [hunk 11]
```diff
@@ -77,44 +80,39 @@ class UserController @Inject()(val messagesApi: MessagesApi)
...
-              "loggedTime" -> result.map {
-                case ((annotation, day), duration) =>
-                  Json.obj(
-                    "annotation" -> annotation,
-                    "day" -> day,
-                    "durationInSeconds" -> duration.toSeconds
-                  )
-              }
-            )
+  def usersLoggedTime = SecuredAction.async(validateJson[TimeSpanRequest]) { implicit request =>
```
Comment (jstriebel) [inline app/controllers/UserController.scala]: +1 for extracting `validateJson`

File: app/controllers/UserController.scala [hunk 12]
```diff
@@ -149,15 +147,15 @@ class UserController @Inject()(val messagesApi: MessagesApi)
   // REST API
   def list = SecuredAction.async { implicit request =>
     UsingFilters(
-      Filter("includeAnonymous", (value: Boolean, el: User) => value || !el.isAnonymous, default = Some("false")),
```
Comment (jstriebel) [inline app/controllers/UserController.scala]: Why is the `includeAnonymous` Filter removed?

File: app/controllers/UserController.scala [hunk 13]
```diff
@@ -190,20 +192,23 @@ class UserController @Inject()(val messagesApi: MessagesApi)
     withJsonBodyUsing(userUpdateReader) {
       case (firstName, lastName, email, isActive, isAdmin, assignedMemberships, experiences) =>
         for {
-          user <- UserDAO.findOneById(userId) ?~> Messages("user.notFound")
```
Comment (jstriebel) [inline app/controllers/UserController.scala]: Why not keep `findOneById`? I think we use it quite often.

Comment (jstriebel) [reply app/controllers/UserController.scala]: Also unpacking with the `Messages("user.notFound")` should be possible in an abstract `findOneById`, maybe having a `match` expression switching the message-identifier.

File: app/controllers/UserController.scala (hunk 10 above)
Comment (fm3) [reply app/controllers/UserController.scala]: good catch, needs `Fox.assertTrue`. The ?~> just adds the failure message

File: app/controllers/UserController.scala (hunk 12 above)
Comment (fm3) [reply app/controllers/UserController.scala]: the backend does not support anonymous users anymore (was for amazon mturk users).

File: app/controllers/UserController.scala (hunk 13 above)
Comment (fm3) [reply app/controllers/UserController.scala]: Not sure I understand you here. UserSQLDAO replaces UserDAO here. The findOne methods everywhere expect an ObjectId, which is why there is no explicit “byId” in the name. The “user.notFound” failure message is added here since the SQLDAOs have more specific database-related failure messages

Comment (jstriebel) [reply app/controllers/UserController.scala]: :+1: , see #2940

File: app/controllers/UserController.scala (hunk 10 above)
Comment (jstriebel) [reply app/controllers/UserController.scala]: I wasn't aware of the implicit conversion, see #2941.

Let's double check all the changes where `user.isEditableBy` was used before, if they have `assertBoolean`/`assertTrue`.

File: app/controllers/UserController.scala (hunk 9 above)
Comment (jstriebel) [reply app/controllers/UserController.scala]: see #2942

File: app/models/annotation/AnnotationService.scala [hunk 14]
```diff
@@ -43,17 +39,20 @@ object AnnotationService
...
-        for {
-          _ <- (user.isTeamManagerInOrg(user.organization) || user.isAdmin)
-          organization <- OrganizationSQLDAO.findOneByName(user.organization)
-          organizationTeamId <- OrganizationSQLDAO.findOrganizationTeamId(organization._id)
-        } yield organizationTeamId
-    }
+  private def selectSuitableTeam(user: UserSQL, dataSet: DataSet)(implicit ctx: DBAccessContext): Fox[ObjectId] = {
+    (for {
+      userTeamIds <- user.teamIds
+    } yield {
```
Comment (jstriebel) [inline app/models/annotation/AnnotationService.scala]: What exactly is `flatten`ed here? Could we use [`async`](https://github.com/scala/scala-async) to circumvent those `(for {…} yield { for {} yield x}).flatten` patterns?

File: app/models/team/Team.scala [hunk 15]
```diff
@@ -138,15 +138,20 @@ object TeamSQLDAO extends SQLDAO[TeamSQL, TeamsRow, Teams] {
...
+  def couldBeAdministratedBy(user: UserSQL) =
+    for {
+      organization <- user.organization
+    } yield organization.name == this.organization
 
-  def isAdminOfOrganization(user: User) =
-    user.organization == organization && user.isAdmin
+  def assertCouldBeAdministratedBy(user: UserSQL) =
+    for {
+      asBoolean <- couldBeAdministratedBy(user)
```
Comment (jstriebel) [inline app/models/team/Team.scala]: This syntax is not very intuitive to me. Having something like the following would be awesome (in my opinion, of course):
```
async def assertCouldBeAdministratedBy(user: UserSQL) = 
  if (await couldBeAdministratedBy(user))
    Success()
  else {
    // not sure with the syntax here, but something like
    throw Messages("notAllowed")
    Failure()
  }
```

a) What is your opinion about this, @fm3 ?
b) Is something similar possible?

Comment (jstriebel) [reply app/models/team/Team.scala]: (having the bool2fox conversion explicitely would help to make it clearer, but I still find an `if` statement appropriate)

Comment (fm3) [reply app/models/team/Team.scala]: should be `Fox.assertTrue(team.couldBeAdministratedBy(user))`. will change.

File: app/models/user/UserService.scala [hunk 16]
```diff
@@ -34,97 +34,109 @@ object UserService extends FoxImplicits with IdentityService[User] {
...
     }
-    UserDAO.update(user._id, firstName, lastName, email, activated, isAdmin, teams, experiences).map {
-      result =>
-        UserCache.invalidateUser(user.id)
-        if(user.email != email)
-          WebknossosSilhouette.environment.tokenDAO.updateEmail(user.email, email)
-        result
-    }
+    for {
+      _ <- UserSQLDAO.updateValues(user._id, firstName, lastName, email, isAdmin, !activated)
```
Comment (jstriebel) [inline app/models/user/UserService.scala]: Why now `!activated`? Probably the attribute name would help here.

File: app/models/team/Team.scala (hunk 15 above)
Comment (jstriebel) [reply app/models/team/Team.scala]: :+1:

File: app/oxalis/thirdparty/BrainTracing.scala [hunk 17]
```diff
@@ -34,41 +34,42 @@ object BrainTracing extends LazyLogging with FoxImplicits {
...
-              Success("braintracing.exists")
-            case _ =>
-              Success("braintracing.error")
-          })
-          logger.trace(s"Creation of account ${user.email} returned Status: ${response.status} Body: ${response.body}")
-        }
-      brainTracingRequest.onFailure{
-        case e: Exception =>
-          logger.error(s"Failed to register user '${user.email}' in brain tracing db. Exception: ${e.getMessage}")
+  def registerIfNeeded(user: UserSQL): Fox[String] =
```
Comment (jstriebel) [inline app/oxalis/thirdparty/BrainTracing.scala]: @fm3 Did you double check to change all `register` calls to `registerIfNeeded`? Also here, maybe just use an inner function, if the new `register` should actually not be called.

File: app/oxalis/thirdparty/BrainTracing.scala [hunk 18]
```diff
@@ -96,57 +97,59 @@ object BrainTracing extends LazyLogging with FoxImplicits {
         a.id))
   }
 
-  def logTime(user: User, time: Long, annotation: Option[AnnotationSQL])(implicit ctx: DBAccessContext): Future[Boolean] = {
+  def logTimeIfNeeded(user: UserSQL, time: Long, annotation: Option[AnnotationSQL])(implicit ctx: DBAccessContext): Fox[Boolean] = {
```
[the remark this answers is not in the document]
Comment (fm3) [reply app/oxalis/thirdparty/BrainTracing.scala]: actually, there are no calls to logTime or to logTimeIfNeeded. We don’t log time to BrainTracing anymore, haven’t been doing that for severall months (instead they fetch time via our API now). I’m removing that functionality in #2939

File: app/oxalis/thirdparty/BrainTracing.scala (hunk 17 above)
Comment (fm3) [reply app/oxalis/thirdparty/BrainTracing.scala]: I made register private, forbidding external calls to it

File: app/models/user/User.scala [hunk 19]
```diff
@@ -41,29 +35,111 @@ case class UserSQL(
...
-        user.isAdmin,
-        user.isSuperUser,
-        !user.isActive,
-        System.currentTimeMillis())
-  }
+      (otherIsTeamManagerOrAdmin || teamMemberships.isEmpty)
+    }
+
+  def isAdminOf(otherUser: UserSQL): Boolean =
+    this._organization == otherUser._organization && this.isAdmin
```
Comment (jstriebel) [inline app/models/user/User.scala]: `isAdminOf(otherUser._organization)` would be the same and more precise

File: app/models/user/User.scala [hunk 20]
```diff
@@ -41,29 +35,111 @@ case class UserSQL(
...
     for {
-      organization <- OrganizationSQLDAO.findOneByName(user.organization)
+      team <- TeamSQLDAO.findOne(_team)(GlobalAccessContext)
+      teamManagerTeamIds <- teamManagerTeamIds
+    } yield (teamManagerTeamIds.contains(_team) || this.isAdmin && this._organization == team._organization)
+
+  def isTeamManagerOrAdminOfOrg(_organization: ObjectId): Fox[Boolean] =
+    for {
+      teamManagerTeamIds <- teamManagerTeamIds
+    } yield teamManagerTeamIds.nonEmpty || this.isAdmin && this._organization == _organization
```
Comment (jstriebel) [inline app/models/user/User.scala]: `yield teamManagerTeamIds.nonEmpty || isAdminOf(_organization)`

Comment (jstriebel) [reply app/models/user/User.scala]: Also, use `isTeamManagerInOrg(_organization)` should be used, so I guess it's just
```
`isTeamManagerInOrg(_organization) || isAdminOf(_organization)
```

File: app/models/user/User.scala [hunk 21]
```diff
@@ -41,29 +35,111 @@ case class UserSQL(
...
+    } yield (otherUserTeamIds.intersect(teamManagerTeamIds).nonEmpty || this.isAdminOf(otherUser))
 
-object UserSQL {
-  def fromUser(user: User)(implicit ctx: DBAccessContext): Fox[UserSQL] =
+  def isTeamManagerOrAdminOf(_team: ObjectId): Fox[Boolean] =
     for {
-      organization <- OrganizationSQLDAO.findOneByName(user.organization)
+      team <- TeamSQLDAO.findOne(_team)(GlobalAccessContext)
+      teamManagerTeamIds <- teamManagerTeamIds
+    } yield (teamManagerTeamIds.contains(_team) || this.isAdmin && this._organization == team._organization)
```
Comment (jstriebel) [inline app/models/user/User.scala]: `… || isAdminOf(team._organization)`

File: app/oxalis/thirdparty/BrainTracing.scala (hunk 18 above)
[the remark this answers is not in the document]
Comment (jstriebel) [reply app/oxalis/thirdparty/BrainTracing.scala]: :+1:

File: app/oxalis/thirdparty/BrainTracing.scala (hunk 17 above)
Comment (jstriebel) [reply app/oxalis/thirdparty/BrainTracing.scala]: True, didn't notice this!

**Comment by fm3** [comment]
Thanks a lot! I will incorporate your feedback!
the negation in !activated was previously done within the UserDAO (which is now gone) – the field name is `isDeactivated` – I will add explicitly add the name in the call.

File: app/controllers/ConfigurationController.scala (hunk 5 above)
Comment (fm3) [reply app/controllers/ConfigurationController.scala]: or not, since this is not team-specific (kind of a special case, all teammanagers are allowed to do this, no matter which team they are in). But I did this replacement in another instance.

**Comment by jstriebel** [comment]
@fm3 : I think the change in `Team.scala` is missing, but besides that LGTM! Let's merge #2939 (on Monday?) first and with that remove `logTime`/`logTimeIfNeeded`. Then this is ready :ship:
